### PR TITLE
Binheap

### PR DESCRIPTION
**Pull request by Jakeand3rson**
I had some fun with the tests, making sure the heap was fully operational. Be aware, there is a test that takes around 30 seconds to pass. Feel free to lower the number if you don't want to wait.

### CODE REVIEW

File: binary_heap.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            child_one_position = child_two_position - 1
+
+            if self.heap_list[child_one_position] >= self.heap_list[child_two_position]:
+                item = self.heap_list[child_one_position]
+                position = child_one_position
+            if self.heap_list[child_two_position] > self.heap_list[child_one_position]:
+                item = self.heap_list[child_two_position]
+                position = child_two_position
+
+        return return_item
```
Comment (cewing) [inline binary_heap.py:49]: don't leave broken things you aren't planning on using in your repo.

File: bin_heap.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python
+
+
+class BinaryHeap(object):
+    def __init__(self):
+        self.heaplist = [0]
+        self.currentsize = 0
```
Comment (cewing) [inline bin_heap.py:7]: I would love to see you work out how to make this work without carrying the weird placeholder to allow 1-based indexing.  For one thing, it's non-pythonic.  For another, it prevents you including values less than 0 (and restricts you to numeric contents, rather than any self-compatible object).  Finally, it forces you to carry an extra instance attribute that should be required (`currentsize`)

File: bin_heap.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    def push(self, val):
+        self.heaplist.append(val)
+        self.currentsize += 1
+        self.bubble_up(self.currentsize)
```
Comment (cewing) [inline bin_heap.py:12]: this is definitely the right way to do this.  A helper method works great for this sort of thing.

File: bin_heap.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    def pop(self):
+        return_val = self.heaplist[1]
+        self.heaplist[1] = self.heaplist[self.currentsize]
+        self.currentsize = self.currentsize - 1
+        self.heaplist.pop()
+        self.settle_down(1)
+        return return_val
```
Comment (cewing) [inline bin_heap.py:20]: This one-based indexing makes my head hurt.

File: bin_heap.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    def bubble_up(self, val):
+        while val // 2 > 0:
+            if self.heaplist[val] < self.heaplist[val // 2]:
+                swappee = self.heaplist[val // 2]
+                self.heaplist[val // 2] = self.heaplist[val]
+                self.heaplist[val] = swappee
+            val = val // 2
```
Comment (cewing) [inline bin_heap.py:28]: just a fun little FYI:

``` python
>>> val = 20
>>> val // 2
10
>>> val >> 1
10
```

`>>` is the Python bit-wise right shift operator.  Shifting any binary value right by 1 is the same as doing a floored division by 2, but is _lots_ faster.

File: bin_heap.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    def settle_down(self, val):
+        while (val * 2) <= self.currentsize:
+            sc = self.youngest(val)
```
Comment (cewing) [inline bin_heap.py:32]: nice helper method here.  Saves having that block of ugly if/else right in this position.

File: bin_heap.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        self.heaplist = [0]
+        self.currentsize = 0
+
+    def push(self, val):
+        self.heaplist.append(val)
+        self.currentsize += 1
+        self.bubble_up(self.currentsize)
+
+    def pop(self):
+        return_val = self.heaplist[1]
```
Comment (cewing) [inline bin_heap.py:15]: When the heap is "empty", this raises an index error, as expected, but it looks odd because the index requested is `1`.  This means that it is clear to the user that although the heap is reporting itself as empty it really has one thing in it.  That's odd.

File: bin_heap.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            else:
+                return val * 2 + 1
+
+    def buildheap(self, alist):
+        x = len(alist) // 2
+        self.currentsize = len(alist)
+        self.heaplist = [0] + alist[:]
+        while (x > 0):
+            self.settle_down(x)
+            x = x - 1
```
Comment (cewing) [inline bin_heap.py:54]: this seems odd to me.  Why not just iterate over `alist` and push the items onto the heap?

Also, you might consider implementing this as a class method so that you don't have to create a heap before calling it:

``` python
# less good
h = BinaryHeap()
h.buildheap([4,12,6,7,3,4,23,1,6,65])

# more good
h = BinaryHeap.buildheap([4,12,6,7,3,4,23,1,6,65])
```

File: test_binary_heap.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    for i in random.sample(range(100), 100):
+        h.push(i)
+    assert all_items(h) == range(0, 100)
+
+
+def test_really_really_big_random_input():
+    h = BinaryHeap()
+    for i in random.sample(range(1000000), 1000000):
+        h.push(i)
+    assert all_items(h) == range(0, 1000000)
```
Comment (cewing) [inline test_binary_heap.py:52]: I'd like to see more tests around popping empty heaps, so on.
